dm/checker/mysql_server.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/pingcap/tidb-tools/pkg/dbutil"
+	"github.com/pingcap/tidb-tools/pkg/utils"
+)
+
+// MySQLVersionChecker checks mysql/mariadb/rds,... version.
+type MySQLVersionChecker struct {
+	db     *sql.DB
+	dbinfo *dbutil.DBConfig
+}
+
+// NewMySQLVersionChecker returns a Checker
+func NewMySQLVersionChecker(db *sql.DB, dbinfo *dbutil.DBConfig) RealChecker {
+	return &MySQLVersionChecker{db: db, dbinfo: dbinfo}
+}
+
+// SupportedVersion defines the MySQL/MariaDB version that DM/syncer supports
+// * 5.6.0 <= MySQL Version
+// * 10.1.2 <= Mariadb Version
+var SupportedVersion = map[string]struct {
+	Min MySQLVersion
+	Max MySQLVersion
+}{
+	"mysql": {
+		MySQLVersion{5, 6, 0},
+		MaxVersion,
+	},
+	"mariadb": {
+		MySQLVersion{10, 1, 2},
+		MaxVersion,
```

Comment:
remember that newer MariaDB doesn't fit DM well. maybe we should be more strict about the version and this check could raise warning not error? (in next PR)

https://github.com/pingcap/dm/issues/2266

Reply:
MariaDB seems to have a very different permission limit than MySQL, I wonder if we can not report MariaDB error.

Or we could just not support MariaDB in pre-check until we sorted out a better MariaDB support scheme.